src/Snafu.md

```diff
@@ -370,6 +371,196 @@ enum Error {
 }
 ```
 
+## Providing data beyond the `Error` trait
+
+<div style="border: 3px solid red;">
+
+**Reviewer Notes**
+
+Please check...
+
+- the usual English prose things: spelling, grammar, flow.
+- How big should the examples be / what should be doc-hidden
+- Is the motivation for the provider API explained enough? Do we need to?
+
+</div>
+
+
+
+When the [`unstable-provider-api` feature
+flag](#unstable-provider-api) is enabled, errors will implement the
+[`Provider` API][provider API]. This allows arbitrary data to be
+associated with an error instance, expanding the abilities of the
+receiver of the error:
+
+```rust,ignore
+use snafu::prelude::*;
+
+#[derive(Debug)]
+struct UserId(u8);
+
+#[derive(Debug, Snafu)]
+struct Error {
+    #[snafu(provide)]
+    user_id: UserId,
+}
+
+let e = Snafu { user_id: UserId(0) }.build();
+let e = &e as &dyn std::error::Error;
+let user_id = e.request_ref::<UserId>().expect("Must have a user id");
+println!("{user_id:?} experienced an error");
+```
+
+### Automatically provided data
+
+By default, `source` and `backtrace` fields are exposed to the
+provider API. Additionally, any data provided by the wrapped error
+will be available on the wrapping error:
+
+```rust,ignore
+use snafu::{prelude::*, IntoError};
+
+#[derive(Debug)]
+struct UserId(u8);
+
+#[derive(Debug, Snafu)]
+struct InnerError {
+    #[snafu(provide)]
+    user_id: UserId,
+    backtrace: snafu::Backtrace,
+}
+
+#[derive(Debug, Snafu)]
+struct OuterError {
+    source: InnerError,
+}
+
+let outer = OuterSnafu.into_error(InnerSnafu { user_id: UserId(0) }.build());
+let outer = &outer as &dyn std::error::Error;
+
+// We can get the source error and downcast it at once
+outer
+    .request_ref::<InnerError>()
+    .expect("Must have a source");
+
+// We can get the deepest backtrace
+outer
+    .request_ref::<InnerError>()
+    .expect("Must have a backtrace");
```

Comment:
I think this piece should use `.request_ref::<snafu::Bracktrace>()` rather than `InnerError`?

Reply:
Thanks! Addressed in [this force-push](https://github.com/shepmaster/snafu/compare/2d827deff4fba3557916bcc20f9a295fef86a5b3..bde44dc2348c80f210312b0bac0ba3f16ae3c169)